File: TemperatureSensor.cpp
#include "TemperatureSensor.h"
#include <iostream>
#include <cstdlib>

TemperatureSensor::TemperatureSensor(const std::string& id, float min, float max)
	: Sensor(id, "Temperature", "∞C"), m_thresholdMin(min), m_thresholdMax(max)
{
}

TemperatureSensor::~TemperatureSensor()
{
	std::cout << "TemperatureSensor destroyed: " << m_id << std::endl;
}

float TemperatureSensor::ReadValue()
{
	// Ёмул€ци€ случайного измерени€ температуры
	m_lastValue = static_cast<float>(rand() % 60 - 10);
	std::cout << "Temperature: " << m_lastValue << " ∞C" << std::endl;
	return m_lastValue;
}

bool TemperatureSensor::IsOutOfRange() const
{
	return (m_lastValue < m_thresholdMin || m_lastValue > m_thresholdMax);
}

std::string TemperatureSensor::GetWarningLevel() const
{
	if (IsOutOfRange()) return "CRITICAL";
	return "NORMAL";
}

File: Lab04/Lab04Monitoring/TemperatureSensor.cs
// Конкретный датчик — температурный.
using System;

namespace Lab04Monitoring
{
    public class TemperatureSensor : Sensor
    {
        // Общий генератор случайных чисел для имитации показаний.
        private static readonly Random _rnd = new Random();

        // Допустимый диапазон температур.
        private readonly float _thresholdMin;
        private readonly float _thresholdMax;

        // Конструктор принимает идентификатор датчика и допустимый диапазон.
        public TemperatureSensor(string id, float min, float max)
            : base(id, $"Temperature sensor {id}", "C")
        {
            _thresholdMin = min;
            _thresholdMax = max;
        }

        // Имитация измерения температуры.
        public override float ReadValue()
        {
            // Случайное значение в диапазоне [-10; 50) градусов.
            LastValue = (float)(_rnd.NextDouble() * 60.0 - 10.0);
            Status = "OK";

            Logger.Instance.Log($"[{Id}] measured {LastValue:F2} {Unit}");
            return LastValue;
        }

        // Проверка, выходит ли измерение за допустимый диапазон.
        public bool IsOutOfRange()
        {
            return LastValue < _thresholdMin || LastValue > _thresholdMax;
        }

        // Текстовое представление "нормально/критично".
        public string GetWarningLevel()
        {
            return IsOutOfRange() ? "CRITICAL" : "NORMAL";
        }
    }
}
